use subtest for table units (pkg/scheduler/util)

xchapter7x opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->
**What this PR does / why we need it**: Update scheduler's unit table tests to use subtest

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

**Special notes for your reviewer**:
breaks up PR: https://github.com/kubernetes/kubernetes/pull/63281
/ref #63267

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
This PR will leverage subtests on the existing table tests for the scheduler units.
Some refactoring of error/status messages and functions to align with new approach.

```

xchapter7x committed:
use subtest for table units

implement PR feedback

  - https://github.com/kubernetes/kubernetes/pull/63710#discussion_r187744182
  - https://github.com/kubernetes/kubernetes/pull/63710#discussion_r187744017

use consistent naming of subtest names

  - also removes a subtest where there is state
  bleeding out from each iteration on the table.

remove whitespace changes

wgliang commented:
/ok-to-test

xchapter7x commented:
/test pull-kubernetes-e2e-kops-aws

xchapter7x commented:
/assign @davidopp

misterikkit commented on pkg/scheduler/util/backoff_utils_test.go:
```diff
@@ -64,23 +64,33 @@ func TestBackoff(t *testing.T) {
-	if duration != 1*time.Second {
-		t.Errorf("expected: 1, got %s", duration.String())
-	}
+	t.Run("Verify that we split on namespaces correctly, same name, different namespace", func(t *testing.T) {
```
I don't think this t.Run improves the clarity of the test. Similarly, the two subtests below might make more sense as better error messages.

xchapter7x replied:
agreed. i meant to come back and revert this bit, as its a bit out of scope for the add subtests on table driven units objective.

misterikkit commented on pkg/scheduler/util/utils_test.go:
```diff
@@ -294,12 +300,14 @@ func TestHostPortInfo_Check(t *testing.T) {
-		if hp.CheckConflict(test.check.ip, test.check.protocol, test.check.port) != test.expect {
-			t.Errorf("%v failed, expected %t; got %t", test.desc, test.expect, !test.expect)
-		}
+		t.Run(test.desc, func(t *testing.T) {
```
rename desc to name?

xchapter7x replied:
👍

k82cn commented:
/assign @misterikkit 
/unassign @davidopp

k82cn commented:
@misterikkit , would you also help to review other PRs about subtest?

misterikkit commented:
> @misterikkit , would you also help to review other PRs about subtest?

I think I replied to them all. Feel free to assign if I missed one. I think someone has to approve after I give LGTM though.

k82cn commented:
great, I can help to approve them :)

misterikkit commented on pkg/scheduler/util/backoff_utils_test.go:
```diff
@@ -64,13 +64,17 @@ func TestBackoff(t *testing.T) {
+				t.Errorf("expected: %s, got %s for %s", test.expectedDuration.String(), duration.String(), test.podID)
+			}
+
+			clock.t = clock.t.Add(test.advanceClock)
```
Hmmm... this is another test that expects the cases to be run in sequence and carries over state.

misterikkit commented on pkg/scheduler/util/utils_test.go:
```diff
@@ -294,12 +300,14 @@ func TestHostPortInfo_Check(t *testing.T) {
+				hp.Add(param.ip, param.protocol, param.port)
+			}
+			if hp.CheckConflict(test.check.ip, test.check.protocol, test.check.port) != test.expect {
+				t.Errorf("%v failed, expected %t; got %t", test.name, test.expect, !test.expect)
```
remove test.name from log

misterikkit commented on pkg/scheduler/util/utils_test.go:
```diff
@@ -190,69 +194,71 @@ func TestHostPortInfo_AddRemove(t *testing.T) {
-			t.Errorf("%v failed: expect length %d; got %d", test.desc, test.length, hp.Len())
-			t.Error(hp)
-		}
+		t.Run(test.desc, func(t *testing.T) {
```
please rename desc to name as you did with the other test.

misterikkit commented on pkg/scheduler/util/utils_test.go:
```diff
@@ -190,69 +194,71 @@ func TestHostPortInfo_AddRemove(t *testing.T) {
+				hp.Remove(param.ip, param.protocol, param.port)
+			}
+			if hp.Len() != test.length {
+				t.Errorf("failed: expect length %d; got %d", test.length, hp.Len())
```
You might want to remove "failed:" from the message.

misterikkit commented on pkg/scheduler/util/utils_test.go:
```diff
@@ -294,12 +300,14 @@ func TestHostPortInfo_Check(t *testing.T) {
+				hp.Add(param.ip, param.protocol, param.port)
+			}
+			if hp.CheckConflict(test.check.ip, test.check.protocol, test.check.port) != test.expect {
+				t.Errorf("failed, expected %t; got %t", test.expect, !test.expect)
```
as above, please remove "failed,"

xchapter7x committed:
remove failed from output

  https://github.com/kubernetes/kubernetes/pull/63710#discussion_r194580752
  https://github.com/kubernetes/kubernetes/pull/63710#discussion_r194580872

xchapter7x commented:
@k82cn @misterikkit

All requested changes/feedback should now be addressed. Just waiting for the checks to go green.

-John
